crates/gpui/src/platform/windows/dispatcher.rs

```diff
@@ -1,74 +1,119 @@
 use std::{
-    sync::{
-        atomic::{AtomicIsize, Ordering},
-        Arc,
-    },
+    ffi::c_void,
     thread::{current, ThreadId},
+    time::Duration,
 };
 
 use async_task::Runnable;
-use flume::Sender;
 use parking::Parker;
 use parking_lot::Mutex;
-use windows::Win32::{Foundation::*, System::Threading::*};
+use util::ResultExt;
+use windows::{
+    Foundation::TimeSpan,
+    System::{
+        DispatcherQueue, DispatcherQueueController, DispatcherQueueHandler,
+        Threading::{
+            ThreadPool, ThreadPoolTimer, TimerElapsedHandler, WorkItemHandler, WorkItemOptions,
+            WorkItemPriority,
+        },
+    },
+    Win32::System::WinRT::{
+        CreateDispatcherQueueController, DispatcherQueueOptions, DQTAT_COM_NONE,
+        DQTYPE_THREAD_CURRENT,
+    },
+};
 
 use crate::{PlatformDispatcher, TaskLabel};
 
+macro_rules! generate_handler {
+    ($handler_type: tt, $runnable: ident, $has_param: ident) => {{
+        macro_rules! generate_func {
```

Comment:
What's going on with this recursive macro definition?

Reply:
Currently, there are two kinds of runners, one that accepts parameters and one that does not. This `generate_func` macro is used to generate either `|_|` (with parameters) or `||` (without parameters).

Runner without param:
```rust
    let task_wrapper = TaskWrapper(runnable.into_raw().as_ptr() as *mut c_void);
    DispatcherQueueHandler::new(move || {
        let task = unsafe {
            let captured = task_wrapper;
            Runnable::<()>::from_raw(std::ptr::NonNull::new_unchecked(captured.0 as *mut ()))
        };
        task.run();
        Ok(())
    })
```
Runner with param:
```rust
    let task_wrapper = TaskWrapper(runnable.into_raw().as_ptr() as *mut c_void);
    WorkItemHandler::new(move |_| {
        let task = unsafe {
            let captured = task_wrapper;
            Runnable::<()>::from_raw(std::ptr::NonNull::new_unchecked(captured.0 as *mut ()))
        };
        task.run();
        Ok(())
    })
```
The only difference between the two is a single character, so the `generate_func` macro is used internally within the macro to assist with this.

The reason why the `generate_func` macro is placed inside `generate_handler` is due to hygiene issues with macros; I need to capture some variables.

Reply:
Ahhh, I see, thank you :)